scrapy_poet/injection.py

```diff
@@ -170,20 +181,48 @@ def build_plan(self, request: Request) -> andi.Plan:
             # Callable[[Callable], Optional[Callable]] but the registry
             # returns the typing for ``dict.get()`` method.
             overrides=self.registry.overrides_for(request.url).get,  # type: ignore[arg-type]
-            custom_builder_fn=self._get_item_builder(request),
+            custom_builder_fn=self._get_custom_builder(request),
         )
 
-    def _get_item_builder(
+    def _get_custom_builder(
         self, request: Request
     ) -> Callable[[Callable], Optional[Callable]]:
         """Return a function suitable for passing as ``custom_builder_fn`` to ``andi.plan``.
 
         The returned function can map an item to a factory for that item based
-        on the registry.
+        on the registry and also supports filling :class:`.DynamicDeps`.
         """
 
         @functools.lru_cache(maxsize=None)  # to minimize the registry queries
         def mapping_fn(item_cls: Callable) -> Optional[Callable]:
+            # building DynamicDeps
+            if item_cls is DynamicDeps:
+                dynamic_types = request.meta.get("inject", [])
+                if not dynamic_types:
+                    return lambda: {}
+
+                # inspired by dataclasses._create_fn()
+                args = [
+                    f"{type_.__name__}_arg: {type_.__name__}" for type_ in dynamic_types
+                ]
+                args_str = ", ".join(args)
+                result_args = [
+                    f"{type_.__name__}: {type_.__name__}_arg" for type_ in dynamic_types
+                ]
+                result_args_str = ", ".join(result_args)
+                ns = {type_.__name__: type_ for type_ in dynamic_types}
+                create_args = ns.keys()
+                create_args_str = ", ".join(create_args)
+                txt = (
+                    f"def __create_fn__({create_args_str}):\n"
+                    f" def dynamic_deps_factory({args_str}) -> DynamicDeps:\n"
+                    f"  return DynamicDeps({{{result_args_str}}})\n"
+                    f" return dynamic_deps_factory"
+                )
+                exec(txt, globals(), ns)
+                return ns["__create_fn__"](*dynamic_types)
```

Comment:
Do you think this code (or a part of it) can be moved to a utility function, and unit tested separately? It looks quite tricky :)

Reply:
Done